Adds VideoFilter class to modify frames

daniel-falk opened the pull request:
Using this class a video object can be initiated with a frame filter and
then be used as a regular vi3o video object. We could e.g. choose to
only use the top half of the image by creating the video with:

video = VideoFilter(fname, lambda frame: frame[:frame.shape[0]//2])

The frame meta data can also be modified from the filter function by
disabling the default copying of meta data:

def offset_time(frame):
    frame.systime += 3600
    return frame

video = VideoFilter(fname, offset_time, copy_frame_meta=False)

This class helps to avoid propagating information about e.g. crop
together with the video through the whole codebase. It is compatible
with SyncedVideos and works with both Mjpg and Mkv video. It can also be
used recursively with itself. It does add some copies of each frame and
should thus be avoided for high frame rate processing if CPU or memory
bandwith is a bottleneck.

Change-Id: I6341ff36151ca744a25223e81990a39a13ba325c

danielfa-axis committed:
Adds VideoFilter class to modify frames

Using this class a video object can be initiated with a frame filter and
then be used as a regular vi3o video object. We could e.g. choose to
only use the top half of the image by creating the video with:

video = VideoFilter(fname, lambda frame: frame[:frame.shape[0]//2])

The frame meta data can also be modified from the filter function:

def offset_time(frame):
    frame.systime += 3600
    return frame

video = VideoFilter(fname, offset_time)

This class helps to avoid propagating information about e.g. crop
together with the video through the whole codebase. It is compatible
with SyncedVideos and works with both Mjpg and Mkv video. It can also be
used recursively with itself.

Change-Id: I6341ff36151ca744a25223e81990a39a13ba325c

hakanardo commented:
Why do you need to copy the data? If you want to support adding metadata with fiters returning ndarrays you can do:

    frame = img.view(Frame)

to convert an ndarray to a Frame object without copying the data. Also, note that slicing a Frame object will result in a new Frame object.

daniel-falk commented:
- Good point, removed the copy.
- Also removed the option to not copy meta data, instead changed the function to only add meta data that is not existing in the frame returned by the filter.

Slicing a Frame results in a new Frame, but all meta data is lost:
```
>>> from vi3o import Video
>>> v = Video("test/t.mjpg")
>>> v[0].systime
1445859308.91
>>> v[0][:].systime
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Frame' object has no attribute 'systime'
```

Is this expected?

daniel-falk commented:
I think the meta-data copy is needed anyways though, since we want it to work with e.g. opencv filters.

hakanardo commented:
Yes unfortunately. Preserving the metadata would be desirable and is
probably possible, but we'll have to dig into the the nupy subcalssing
machinery:

    https://docs.scipy.org/doc/numpy/user/basics.subclassing.html

On Sat, Feb 29, 2020 at 12:30 PM Daniel Falk <notifications@github.com>
wrote:

>
>    - Good point, removed the copy.
>    - Also removed the option to not copy meta data, instead changed the
>    function to only add meta data that is not existing in the frame returned
>    by the filter.
>
> Slicing a Frame results in a new Frame, but all meta data is lost:
> `>>> from vi3o import Video
>
> v = Video("test/t.mjpg")
> v[0].systime
> 1445859308.91
> v[0][:].systime
> Traceback (most recent call last):
> File "", line 1, in
> AttributeError: 'Frame' object has no attribute 'systime'`
>
> Is this expected?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hakanardo/vi3o/pull/21?email_source=notifications&email_token=AA2GZ4JQSJSLG6BNJWYX5DLRFDYWRA5CNFSM4K3AA6MKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENLXXDA#issuecomment-592935820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2GZ4KY7QQIRE4GER67GXTRFDYWRANCNFSM4K3AA6MA>
> .
>


-- 
Håkan Ardö

hakanardo commented:
Yes, copying metadata is fine, but the current implementation copes the
image data aswell which is not a good idea.

Maybe we could get rid of the copy_meta_data parameter by only copying
metadata not present in the frame/ndarray returned by the filter?

On Sat, Feb 29, 2020 at 12:33 PM Daniel Falk <notifications@github.com>
wrote:

> I think the meta-data copy is needed anyways though, since we want it to
> work with e.g. opencv filters.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/hakanardo/vi3o/pull/21?email_source=notifications&email_token=AA2GZ4LXGQ2FGYU35HKXMBLRFDY7LA5CNFSM4K3AA6MKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENLXYUY#issuecomment-592936019>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2GZ4JIIWUUV3KTC2UW7GTRFDY7LANCNFSM4K3AA6MA>
> .
>


-- 
Håkan Ardö

daniel-falk commented:
The last patch set did remove the copy of the data altogether and it also removed the copy_meta_data parameter.
